File: datasets/implementations/payment_json.js
/* Local payment-event loader. Outcomes are removed from scoring events. */
(function(global){
  'use strict';const registry=typeof module!=='undefined'?require('../../shared/runtime/dataset-registry'):global.FraudDatasets;
  function load(input){
    const raw=typeof input==='string'?JSON.parse(input):input;
    if(!raw||typeof raw!=='object'||(raw.schema&&raw.schema!=='payment-events/v1'))throw Error('Expected a payment-events/v1 document.');
    const units=raw.units||{time:'minutes',currency:'EUR'};
    if(units.time!=='minutes'||units.currency!=='EUR')throw Error('Convert payment data to minutes and EUR before importing.');
    if(!Array.isArray(raw.accounts)||!raw.accounts.length||!Array.isArray(raw.events)||!raw.events.length)throw Error('A dataset needs nonempty accounts and events.');
    if(raw.accounts.length>256||raw.events.length>20000)throw Error('The browser replay supports at most 256 accounts and 20,000 events. Use a bounded event subset or the Python experiment runner.');
    const external=new Set(),accounts=raw.accounts.map((account,index)=>{if(account.id!==index)throw Error('Account IDs must be contiguous indices; normalize external IDs with the CSV loader.');const id=String(account.external_id??index);if(external.has(id))throw Error('Duplicate external account ID.');external.add(id);return {...account,external_id:id,name:String(account.name??id)};});
    const ids=new Set(),payments=new Set();let last=-Infinity;
    const events=raw.events.map(rawEvent=>{
      const {id,kind,t,u,v,amount}=rawEvent;
      if(typeof id!=='string'||!id||ids.has(id))throw Error('Event IDs must be nonempty and unique.');ids.add(id);
      if(!Number.isFinite(t)||t<0||t<last||!Number.isFinite(amount)||amount<0)throw Error('Events must have finite nonnegative amounts and chronological times.');last=t;
      if(!Number.isInteger(u)||!Number.isInteger(v)||v<0||v>=accounts.length||u< -1||u>=accounts.length||u===v)throw Error('Invalid account index.');
      if(!['payment','deposit','report'].includes(kind)||(kind==='payment')!==(u>=0))throw Error('Invalid event kind/sender combination.');
      if(kind==='payment')payments.add(id);
      const event={id,kind,t,u,v,amount};
      if(rawEvent.settled!==undefined){if(typeof rawEvent.settled!=='boolean')throw Error('settled must be boolean.');event.settled=rawEvent.settled;}
      if(rawEvent.reference!==undefined)event.reference=String(rawEvent.reference);
      return event;
    });
    const truth={};for(const [id,value]of Object.entries(raw.truth||{})){if(!payments.has(id))throw Error('Outcome refers to an unknown payment: '+id);if(value===true||value===1)truth[id]=true;else if(value===false||value===0)truth[id]=false;else if(![null,-1,'unknown',''].includes(value))throw Error('Outcomes must be 0, 1 or unknown.');}
    return {schema:'payment-events/v1',units:{...units},name:String(raw.name||'Imported payments'),size:'imported',seed:null,accounts,events,truth,focus:[],bookmarks:[],startIndex:0,description:String(raw.description||'User-supplied payment events.'),provenance:JSON.parse(JSON.stringify(raw.provenance||{origin:'user-supplied'}))};
  }
  registry.register('payment_json',{load});if(typeof module!=='undefined')module.exports={load};
})(globalThis);

File: datasets/implementations/payment_json.py
"""Validate the common payment-event document used by the offline browser tools."""
import copy, json
from framework.contracts import EventDataset
from ._validation import source, provenance, number, label


def normalize(raw, metadata=None):
    if not isinstance(raw, dict):
        raise ValueError("Expected a payment-event JSON object.")
    if not isinstance(raw.get("truth", {}), dict):
        raise ValueError("truth must map payment IDs to outcomes.")
    if raw.get("schema", "payment-events/v1") != "payment-events/v1":
        raise ValueError("Unsupported payment dataset schema.")
    units = raw.get("units", {"time": "minutes", "currency": "EUR"})
    if units.get("time") != "minutes" or units.get("currency") != "EUR":
        raise ValueError(
            "Browser payment datasets require minutes and EUR; convert explicitly in a dataset loader."
        )
    accounts = raw.get("accounts")
    events = raw.get("events")
    if (
        not isinstance(accounts, list)
        or not accounts
        or not isinstance(events, list)
        or not events
    ):
        raise ValueError("Payment data needs nonempty accounts and events.")
    normalized_accounts = []
    external = set()
    for index, account in enumerate(accounts):
        if account.get("id") != index:
            raise ValueError(
                "Account IDs must be contiguous indices; use the CSV loader to map external IDs."
            )
        external_id = str(account.get("external_id", account["id"]))
        if external_id in external:
            raise ValueError("Duplicate external account identity.")
        external.add(external_id)
        normalized_accounts.append(
            {
                **account,
                "external_id": external_id,
                "name": str(account.get("name", external_id)),
            }
        )
    normalized = []
    ids = set()
    payment_ids = set()
    last = -float("inf")
    for raw_event in events:
        event = {
            key: raw_event[key]
            for key in ("id", "kind", "t", "u", "v", "amount")
            if key in raw_event
        }
        if set(event) != {"id", "kind", "t", "u", "v", "amount"}:
            raise ValueError("Every event requires id, kind, t, u, v and amount.")
        if not isinstance(event["id"], str) or not event["id"] or event["id"] in ids:
            raise ValueError("Event IDs must be nonempty unique strings.")
        ids.add(event["id"])
        event["t"] = number(event["t"], "event time")
        event["amount"] = number(event["amount"], "amount")
        if event["t"] < 0 or event["t"] < last or event["amount"] < 0:
            raise ValueError(
                "Events must be chronological, with nonnegative times and amounts."
            )
        last = event["t"]
        u, v = event["u"], event["v"]
        if (
            type(u) != int
            or type(v) != int
            or not 0 <= v < len(accounts)
            or not -1 <= u < len(accounts)
            or u == v
        ):
            raise ValueError("Invalid event account index.")
        if event["kind"] not in ("payment", "deposit", "report") or (
            event["kind"] == "payment"
        ) != (u >= 0):
            raise ValueError("Invalid event kind/sender combination.")
        if event["kind"] == "payment":
            payment_ids.add(event["id"])
        if "settled" in raw_event:
            if type(raw_event["settled"]) != bool:
                raise ValueError("settled must be boolean.")
            event["settled"] = raw_event["settled"]
        if "reference" in raw_event:
            event["reference"] = str(raw_event["reference"])
        normalized.append(event)
    truth = {}
    for identifier, outcome in raw.get("truth", {}).items():
        if identifier not in payment_ids:
            raise ValueError("Outcome refers to an unknown payment: " + identifier)
        value = label(outcome)
        if value >= 0:
            truth[identifier] = bool(value)
    availability = raw.get("label_available_at", {})
    if not isinstance(availability, dict):
        raise ValueError(
            "label_available_at must map payment IDs to confirmation times in minutes."
        )
    payment_times = {
        event["id"]: event["t"] for event in normalized if event["kind"] == "payment"
    }
    confirmed = {}
    for identifier, value in availability.items():
        if identifier not in truth:
            raise ValueError(
                "Label availability requires a known payment outcome: " + identifier
            )
        value = number(value, "label availability time")
        if value < payment_times[identifier]:
            raise ValueError(
                "A payment outcome cannot be confirmed before the payment."
            )
        confirmed[identifier] = value
    return {
        "schema": "payment-events/v1",
        "units": dict(units),
        "name": str(raw.get("name", "Imported payments")),
        "size": str(raw.get("size", "imported")),
        "seed": raw.get("seed"),
        "accounts": normalized_accounts,
        "events": normalized,
        "truth": truth,
        "focus": [],
        "bookmarks": [],
        "startIndex": 0,
        "description": str(raw.get("description", "User-supplied payment events.")),
        "provenance": copy.deepcopy(
            metadata or raw.get("provenance", {"origin": "user-supplied"})
        ),
        **({"label_available_at": confirmed} if availability else {}),
    }


def load(config, base):
    path = source(config, base)
    metadata = provenance(config, path)
    raw = json.loads(path.read_text())
    if isinstance(raw, dict) and isinstance(raw.get("provenance"), dict):
        metadata["source_provenance"] = copy.deepcopy(raw["provenance"])
    document = normalize(raw, metadata)
    return EventDataset(document, metadata)
